### PR TITLE
Remove libp2p handlers when ConnectionHandler, DHT, and DecentralizedAverager are shut down

### PR DESCRIPTION
**Pull request by borzunov**

### CODE REVIEW

**Review by justheuristic — APPROVED** [review]

File: hivemind/p2p/p2p_daemon.py [hunk 1]
```diff
@@ -553,6 +566,9 @@ async def add_binary_stream_handler(
             self._start_listening()
         await self._client.stream_handler(name, handler, balanced)
 
+    async def remove_binary_stream_handler(self, name: str) -> None:
+        await self._client.remove_stream_handler(name)
+
```
[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/p2p/p2p_daemon.py:571]: It's a private field.

[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/p2p/p2p_daemon.py:571]: `_client.remove_stream_handler` is a private field, and we'd like to keep the public interface for working with binary stream handlers as well (even though it's not currently used).

[the remark this answers is not in the document]
Comment (GreenFatGuy) [reply hivemind/p2p/p2p_daemon.py:571]: Wouldn't it be more consistent to create such method for removing unary handler then? What triggered me is that one type of handlers have public interface and another dont

[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/p2p/p2p_daemon.py:571]: The current logic is the following.

From the public interface's point of view, there are only two types of handlers - binary stream handlers and protobuf handlers. A user is expected to use one of them depending on whether they need serialization.

However, internally, protobuf handlers may work either over unary handlers (if they accept and return just one protobuf), or over binary stream handlers (if they accept or return multiple protobufs). Here, the unary handlers are just an optimization for one edge case (we don't want to open extra connections if there're only one input and one output). This optimization is invisible to a user, i.e., a user just chooses what they want, and `hivemind` chooses the most efficient way to implement that by itself.

**Review by GreenFatGuy — APPROVED** [review]

**Pull request merged**